MAINT: stats: shift `NumericalInverseHermite` to UNU.RAN

tirthasheshpatel opened the pull request:
#### Reference issue

#14600 
Closes gh-14875

#### What does this implement/fix?

Begin translation of FNI to UNU.RAN. Minimal changes have been done to the existing docs, tests, and benchmarks. What's left:

- [x] Tutorial
- [x] Stronger tests
- [x] More docs
- [x] Doc deprecate `tol` parameter
- [x] Add type hints

Two tests with SciPy's distribution suite are failing: one with round-off numerical errors warning (`kappa4` distribution) and the other with an error that CDF was less than 0 (`vonmises` distribution).

A `get_qrng` method has been added to the `_URNG` class to add support for the `qrvs` method of `NumericalInverseHermite`. It samples an array of quasi-uniform random numbers and stores them in a  thread-safe array which is accessed by UNU.RAN while transforming those samples to actual samples using the approximated PPF. As UNU.RAN doesn't provide the truncated boundaries of the distribution used by the generator, we can use the `ppf` method directly (as previously done) to transform the samples.

Backward incompatible changes:

- An array of midpoint u-error was previously calculated and stored in the attribute `fni.midpoint_error`. But there is no way to access the midpoint u-error in UNU.RAN so the attribute has been removed. A `u_error` method has been added instead. Cross-ref gh-14875

#### Additional information
<!--Any additional information you think is important.-->

tirthasheshpatel commented:
@mdhaber Here's the summary of what I have changed and why:

- Removed the docs that emphasized the speed of the method: I think this is better done in the main tutorial where we can explain how UNU.RAN methods are faster than the default method in SciPy with examples like you wrote here.
- Removed the `midpoint_error` attribute because there doesn't seem to be a way to get that using UNU.RAN: I don't this is a major backward-incompatible change.
- Doc-deprecated `tol` parameter: The constructor accepts a `tol` parameter but it has been excluded from the stub file and docs.
- Two tests with SciPy's distribution suite are failing and have been excluded: one with round-off numerical errors warning (`kappa4` distribution) and the other with an error that CDF was less than 0 (`vonmises` distribution).

These changes are pretty much ready for review now. So, let me know if they look good to you.

mdhaber commented:
You mentioned at the top that you would add stronger tests. Conceptually, what are those?
Haven't checked - if both `tol` and `u_error` are provided, what happens?

tirthasheshpatel commented:
> Conceptually, what are those?

I meant adding tests for the `ppf` method and tests from UNU.RAN's test suite both of which have been done.

> Haven't checked - if both `tol` and `u_error` are provided, what happens?

I think you mean `u_resolution`. In that case, `u_resolution` is preferred over `tol`.

mdhaber commented:
I understand the addition of tests that were already in UNU.RAN's test suite, especially if they are specific to the new options. 

> I meant adding tests for the ppf method

The relative error and the u-error of the ppf method of the fast numerical inverse object was tested against the PPF of the distribution (for every distribution not in an exception list) in lines 7199-7206 of `scipy/stats/tests/test_stats.py`. What has changed about the testing of the ppf method?

tirthasheshpatel commented:
> What has changed about the testing of the ppf method?

Just testing with some edge cases and invalid values.

```
    u = [
        # check the correctness of the PPF for equidistant points between
        # 0.02 and 0.98.
        np.linspace(0., 1., num=10000),
        # test the PPF method for empty arrays
        [], [[]],
        # test if nans and infs return nan result.
        [np.nan], [-np.inf, np.nan, np.inf],
        # test if a scalar is returned for a scalar input.
        0,
        # test for arrays with nans, values greater than 1 and less than 0,
        # and some valid values.
        [[np.nan, 0.5, 0.1], [0.2, 0.4, np.inf], [-2, 3, 4]]
    ]
```

mdhaber commented:
> Removed the docs that emphasized the speed of the method:

I tried that removed code out and noticed that generating the `NumericalInverseHermite` object is much slower in that example.

This part:
```
    >>> def time_me():
    ...     fni = NumericalInverseHermite(dist)
    ...     fni.ppf(p)
    >>> time_once(time_me)
```
I got `520.679 ms` with the new code and `8.8897 ms` in master. The time to evaluate the ppf and the maximum error (with defalt settings for both) seem comparable. Is this typical? Have you compared them with the `NumericalInverseHermite` benchmark?

> Removed the midpoint_error attribute because there doesn't seem to be a way to get that using UNU.RAN:

Hmm another lesson against public attributes. But ideally we wouldn't remove that suddenly; it should get a deprecation cycle. Is there no way to access the grid and manually evaluate the error at the midpoints to put in that attribute?

mdhaber commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -1370,6 +1401,510 @@ cdef class NumericalInversePolynomial(Method):
+          `dist`, it is used to set the domain of the distribution.
+        * Otherwise the support is assumed to be :math:`(-\infty, \infty)`.
+
+    order : int, optional
```
What is the default? Consider giving the set with the default listed first rather than specifying `int`.

tirthasheshpatel replied:
Done.

tirthasheshpatel commented:
> I got `520.679 ms` with the new code and `8.8897 ms` in master. The time to evaluate the ppf and the maximum error (with defalt settings for both) seem comparable. Is this typical? Have you compared them with the `NumericalInverseHermite` benchmark?

> Are the calls to the distribution `cdf` method vectorized? I ask because I am wondering if we are incurring `scipy.stats_rv_continuous` overhead in many calls to the method.

Setting up with distributions inheriting from `rv_continuous` is much slower because of extensive input validations and expensive broadcasting and masking done by the methods of the base class. UNU.RAN doesn't call the methods of the distribution in a vectorized manner and hence the slow down. This is kind of a bummer right now but I don't see an easy way to fix this. One ugly hack is to use private methods like `_pdf`, `_cdf` of the distribution. This should work as expected because we know that UNU.RAN always calls these methods with a scalar within the domain (so no need for broadcasting or masking). A long-term solution might be to rewrite UNU.RAN generators with their C source as a reference as mentioned by @rkern [here](https://mail.python.org/pipermail/scipy-dev/2021-April/024701.html):

> The worst-case scenario is that we simply implement the algorithms
> ourselves using the C code as a reference. Given how we want to interact
> with scipy.stats distribution objects and numpy BitGenerators, it's
> entirely possible that's where we want to end up, anyways.

> I noticed that this branch's `NumericalInverseHermite` is _much_ faster with a custom distribution object that doesn't have the overhead

Yes, setting up with distributions that don't inherit from the `rv_continuous` base class is as fast as that on master.

> Is there no way to access the grid and manually evaluate the error at the midpoints to put in that attribute?

Unfortunately, UNU.RAN doesn't provide any functions to inspect the grid. We can patch UNU.RAN but I am against doing that because 

> Also, with that example, `fni.u_error()` took a few seconds. Why does it take so long?

UNU.RAN calls the `cdf` method to evaluate the exact CDF of the distribution in order to calculate the u-error. With `rv_continuous` distributions, this takes a long time but for custom distributions, it should be faster.

mdhaber commented:
> Yes, setting up with distributions that don't inherit from the rv_continuous base class is as fast as that on master.

Well, not in my tests. It still takes 5~8 times as long for me. It looks like sampling might be faster (I guess because `CubicHermiteSpline` is slower), but I'll want to check on a different computer.

<details>

```
import numpy as np
from timeit import timeit
time_once = lambda f: f"{timeit(f, number=1)*1000:.6} ms"

from scipy.stats import NumericalInverseHermite

from scipy.special import ndtr, ndtri

class MyNormal:

    def pdf(self, x):
       return 1/np.sqrt(2*np.pi) * np.exp(-x**2 / 2)

    def cdf(self, x):
       return ndtr(x)

    def ppf(self, x):
       return ndtri(x)

fni = NumericalInverseHermite(MyNormal())

def time_me():
    fni.rvs(100000)

print(time_once(time_me))
````

</details>

tirthasheshpatel commented:
> Well, not in my tests. It still takes 5~8 times as long for me.

Ah, yes. In my tests, I used `math.erf` instead of `ndtr`. So, the setup was almost comparable to the performance on master. Vectorization inevitably introduces some overhead, so the setup takes longer when vectorized methods are used.

<details>

```python
In [1]: from scipy.stats import NumericalInverseHermite

In [2]: from scipy.special import ndtr

In [3]: from math import exp, pi, sqrt, erf

In [4]: import numpy as np

In [5]: class StandardNormal:
   ...:     def pdf(self, x):
   ...:         return 1./sqrt(2.*pi) * exp(-0.5 * x*x)
   ...:     def dpdf(self, x):
   ...:         return 1./sqrt(2.*pi) * -x * exp(-0.5 * x*x)
   ...:     def cdf(self, x):
   ...:         return (1.0 + erf(x / sqrt(2.0))) / 2.0
   ...: 

In [6]: dist = StandardNormal()

In [7]: %timeit NumericalInverseHermite(dist)
11 ms ± 22.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [8]: from scipy.stats._rvs_sampling import NumericalInverseHermite as HINV_Old

In [9]: class StandardNormal:
    ...:     def pdf(self, x):
    ...:         return 1./np.sqrt(2.*pi) * np.exp(-0.5 * x*x)
    ...:     def dpdf(self, x):
    ...:         return 1./np.sqrt(2.*pi) * -x * np.exp(-0.5 * x*x)
    ...:     def cdf(self, x):
    ...:         return ndtr(x)
    ...:     def ppf(self, x):
    ...:         return ndtri(x)
    ...: 

In [10]: from scipy.special import ndtri

In [11]: dist = StandardNormal()

In [12]: %timeit HINV_Old(dist)
9.39 ms ± 706 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [13]: %timeit NumericalInverseHermite(dist)
35.9 ms ± 2.94 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

</details>

chrisb83 commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -1370,6 +1401,526 @@ cdef class NumericalInversePolynomial(Method):
+            raise ValueError("`order` must be either 1, 3, or 5.")
+        if int(max_intervals) != max_intervals or max_intervals <= 1:
+            raise ValueError("`max_intervals' must be an integer greater than 1.")
+        if tol is not None:
```
If both tol and u_resolution are provided, should a warning be raised that u_resolution will be considered and tol should not be used?

chrisb83 replied:
OK, I think you removed `tol` from the documented API and ensured that the value can still be provided. I think we might need to add a deprecation warning but not sure about the process

tirthasheshpatel replied:
> OK, I think you removed `tol` from the documented API and ensured that the value can still be provided. I think we might need to add a deprecation warning but not sure about the process

I haven't raised a deprecation warning because this is a doc-deprecation. But we can probably warn the users in the docs by saying that the constructor also accepts `tol` which is equivalent to `u_resolution` but `u_resolution` should be preferred over `tol` and both the arguments should not be passed at the same time.

tirthasheshpatel replied:
> but it seems to me like the default is to fall back on `tol`.

Yes, I noticed this a couple of days ago and had a chat with @chrisb83. The problem is that `u_resolution` has a non-none default value and there is no way to know when the user passed both `tol` and `u_resolution` i.e. there is no way to know the difference between this:

```py
>>> rng = NumericalInverseHermite(dist, tol=1e-12)
>>> rng = NumericalInverseHermite(dist, u_resolution=1e-10, tol=1e-12)
>>> # in the latter case, did the user pass both u_resolution and tol?
>>> # If so, how to tell that as u_resolution already has a default value of 1e-10
```

One way to deal with this is to keep `u_resolution=None` in the constructor and assign it the value `1e-10` if the user hasn't passed anything. This way, if both `u_resolution` and `tol` are not none, we could raise a warning and prefer `u_resolution` over `tol`. Does this make sense?

swallan replied:
I see the issue that you're dealing with, I'm not aware of a clean way to tell if the keyword default was supplied or not (especially since this is an immutable type).  Setting the default to `None` for `u-resolution` feels like an awkward solution for a new keyword. It could also cause confusion -- similar to how the behavior when `axis=None` in some functions is drastically different than when it is numerically set. 

The folks who will be using the `u_res` parameter will be future implementations, and they will know (or at least be able to find out if you raise a warning) that `tol` is being deprecated if they try to use it. Maintaining backwards compatibility should be prioritized, so I think there is only one real way forward, which is to use `tol` if it has been set. 

the two current usages for `tol` seem to be: 
1. Any previous users that did not pass in a value for `tol` will see the functionality maintained. The `u res` default value matches the previous `tol` value so this already works well.
2. users who did pass in a `tol ` (default `1e-12`  or other value) need to still have the same results regardless of what `u-res` is. 

Trying to force `u-res` to be preferred over `tol` when both are provided seems overly complicated for a scenario where the user is not reading the documentation. I think there would be a smoother path to being merged into main if backwards compatibility is maintained as closely as possible. 

In the end, it seems to me that if `tol` is supplied, you should indicate in a warning that while its input will be honored, it has been replaced with `u_resolution` in scipy 1.x, and it will may be fully deprecated in future releases. 

Others may have different opinions, though!

chrisb83 replied:
> In the end, it seems to me that if tol is supplied, you should indicate in a warning that while its input will be honored, it has been replaced with u_resolution in scipy 1.x, and it will may be fully deprecated in future releases.

I agree, this seems to be the best option we have.

tirthasheshpatel replied:
Thanks for the detailed reply, @swallan!

> The folks who will be using the `u_res` parameter will be future implementations, and they will know (or at least be able to find out if you raise a warning) that `tol` is being deprecated if they try to use it. Maintaining backwards compatibility should be prioritized, so I think there is only one real way forward, which is to use `tol` if it has been set.

Nice catch! I didn't see it this way.

> Trying to force `u-res` to be preferred over `tol` when both are provided seems overly complicated for a scenario where the user is not reading the documentation. I think there would be a smoother path to being merged into main if backwards compatibility is maintained as closely as possible.

I agree. But I wanted to change the default value of `u_resolution` to `1e-10` instead of `1e-12` because UNU.RAN has that as default. This should not break backward compatibility because if the setup succeeded for `u_resolution=1e-12`, it should also succeed for `u_resolution=1e-10`. Moreover, `1e-10` is sufficiently close to machine precision and the setup is faster. This change doesn't affect the user so it seems reasonable to me. Do you agree?

> In the end, it seems to me that if `tol` is supplied, you should indicate in a warning that while its input will be honored, it has been replaced with `u_resolution` in scipy 1.x, and it will may be fully deprecated in future releases

This seems reasonable to me. I will raise a deprecation warning if `tol` is supplied.

swallan replied:
> This change doesn't affect the user 

I made a quick comparison to examine the average difference for these two scenarios:
- previous implementation => new implementation both at 1e-12
- previous implementation @ 1e-12 => new implementation at 1e-10

and then found the bootstrapped confidence intervals for mean absolute error for both:
```
stats.bootstrap([np.abs(diff_12_to_10)], np.mean, confidence_level=.99)
Out[14]: BootstrapResult(confidence_interval=ConfidenceInterval(low=1.448447252711894e-10, high=1.1200871877506022e-09), standard_error=1.630491764207449e-10)

stats.bootstrap([np.abs(diff_both_12)], np.mean, confidence_level=.99)
Out[15]: BootstrapResult(confidence_interval=ConfidenceInterval(low=1.0504267543195413e-12, high=5.4826531970922883e-11), standard_error=8.617733140441905e-12)
```

So the difference between going from previous code @ 1e-12 to new code at 1e-10 has a slightly larger absolute error between the samples. Small, but possibly something to think about. I'm not completely familiar if the results are supposed to match up exactly, so if they are not, disregard.

tirthasheshpatel replied:
Thanks for the analysis!

> So the difference between going from previous code @ 1e-12 to new code at 1e-10 has a slightly larger absolute error between the samples than going from previous to new with both at 1e-12. Small, but possibly something to think about.

I think this is expected because we are moving to a higher error bound. I don't think this change should affect the user very much as it is too small. What do you think @chrisb83?

chrisb83 replied:
Even if we keep the error at 1e-12, there will be a little bit of variation in the random variates due to the change from the Python implementation to UNU.RAN (Samuel's analysis shows this). Changing the u-resolution from 1e-12 to 1e-10 will add a bit of noise again (to isolate the effect, it would be best to compare the same generator with the different values of `u_error`)

From a practical point of view, this should be irrelevant: if you use the simulated rvs in a simulation, differences of order 1e-9 in the underlying rvs should not matter at all.

tirthasheshpatel replied:
Thanks @chrisb83 and @swallan! I have fixed the direction warnings and added one for `tol`. Let me know if there are any other points left to address.

chrisb83 commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -1370,6 +1401,526 @@ cdef class NumericalInversePolynomial(Method):
+        Parameters
+        ----------
+        u : array_like
+            Qunatiles.
```
typo

tirthasheshpatel replied:
done.

chrisb83 commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -1370,6 +1401,526 @@ cdef class NumericalInversePolynomial(Method):
+        max_error : float
+            Maximum u-error.
+        mean_absolute_error : float
+            Maximum absolute u-error.
```
Mean

tirthasheshpatel replied:
done.

chrisb83 commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -1370,6 +1401,526 @@ cdef class NumericalInversePolynomial(Method):
+
+    def qrvs(self, size=None, d=None, qmc_engine=None):
+        """
+        Quasi-random variates of the given RV.
```
signature probably needs to be added to the docstring as for the other method

tirthasheshpatel replied:
done.

chrisb83 reviewed (commented):
I just had a quick look at the code. This looks very comprehensive given that it builds on the existing documentation / testing.

chrisb83 commented:
> An array of midpoint u-error was previously calculated and stored in the attribute fni.midpoint_error. But there is no way to access the midpoint u-error in UNU.RAN so the attribute has been removed. A u_error method has been added instead.

Not sure what the right approach is in that case. Maybe set `midpoint_error` to nan for now and we document that the value is no longer available and will be deprecated?

tirthasheshpatel commented:
@chrisb83 I have added the tutorial. I just copy-pasted the API docs for now as they were elaborate enough.

swallan commented on doc/source/tutorial/stats/sampling_hinv.rst:
```diff
@@ -0,0 +1,191 @@
+    >>> rng2.u_error(sample_size=1000000)
+    UError(max_error=9.32027892364129e-14, mean_absolute_error=1.5194172675685075e-14)
+
+Note that this comes at the cost of increased setup time and number of intervals.
```
Is this entirely correct? This phrasing implies that reducing the u-error by decreasing the u-resolution comes at the cost of the number of intervals, but they in fact increase, which seems to be a positive factor in reducing the u-error. 

I think you may have meant that 
```suggestion
Note that this comes at the cost of computation time as a result of the increased setup time and number of intervals. 
```

tirthasheshpatel replied:
> Is this entirely correct? This phrasing implies that reducing the u-error by decreasing the u-resolution comes at the cost of the number of intervals, but they in fact increase, which seems to be a positive factor in reducing the u-error.

Ah, yes. This makes sense.

> I think you may have meant that

Yes, that's exactly what I meant!

tirthasheshpatel replied:
Changed it this way, thanks!

swallan reviewed (commented):
I took a close look at the documentation, and it looks very thorough. There were a few items that felt a bit unclear, so I left a note on them.

tirthasheshpatel commented on doc/source/tutorial/stats/sampling_hinv.rst:
```diff
@@ -0,0 +1,191 @@
+
+.. plot::
+
+    >>> import numpy as np
```
I need to reimport everything and rewrite the code from scratch again when using the `.. plot::` directive. One way to avoid this could be to put all the code below the `.. plot::` directive in a file under `doc/source/tutorial/stats/plots` and include the resulting plot in the tutorial without showing the code:

    .. plot:: tutorial/stats/plots/hinv_plot.py
    :align: center
    :include-source: 0

swallan replied:
Didn't know this yet! Thanks.

tirthasheshpatel replied:
I have changed it this way. Let's see how it looks.

tirthasheshpatel commented on doc/source/tutorial/stats/sampling_hinv.rst:
```diff
@@ -0,0 +1,191 @@
+    >>> from scipy.stats import norm, genexpon
+    >>> from scipy.special import ndtr
+    >>> import matplotlib.pyplot as plt
+    >>> class StandardNormal:
+    ...     def pdf(self, x):
+    ...        return 1/np.sqrt(2*np.pi) * np.exp(-x**2 / 2)
+    ...     def cdf(self, x):
+    ...        return ndtr(x)
```
Yes, because this is under the `.. plot::` directive

tirthasheshpatel commented on doc/source/tutorial/stats/sampling_hinv.rst:
```diff
@@ -0,0 +1,191 @@
+
+The method is not exact, as it only produces random variates of the approximated
+distribution. Nevertheless, the maximal numerical error in "u-direction" (i.e.
+``|U - CDF(X)|`` where ``X`` is the approximate percentile corressponding to the
```
Corrected both here and in API docs.

tirthasheshpatel commented on doc/source/tutorial/stats/sampling_hinv.rst:
```diff
@@ -0,0 +1,191 @@
+    >>> timeit(f, number=1)
+    0.08671202100003939  # may vary
+
+Please see [1]_ and [2]_ for more details on this method.
```
Done.

tirthasheshpatel commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -1370,6 +1406,538 @@ cdef class NumericalInversePolynomial(Method):
+
+    Given the derivative of the PDF w.r.t the variate (i.e. ``x``), we can use
+    quintic Hermite interpolation to approximate the PPF by passing the ``order``
+    parameter:
```
Done.

tirthasheshpatel commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -1370,6 +1406,538 @@ cdef class NumericalInversePolynomial(Method):
+
+    The u-error can be estimated by calling the `u_error` method. It runs a small
+    Monte-Carlo simulation to estimate the u-error. By default, 100000 samples are
+    used. To change this, you can pass the number of samples as an argument:
```
Done.

tirthasheshpatel commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -1370,6 +1406,538 @@ cdef class NumericalInversePolynomial(Method):
+
+    The method is not exact, as it only produces random variates of the approximated
+    distribution. Nevertheless, the maximal numerical error in "u-direction" (i.e.
+    ``|U - CDF(X)|`` where ``X`` is the approximate percentile corressponding to the
```
Done.

tirthasheshpatel commented on doc/source/tutorial/stats/sampling_hinv.rst:
```diff
@@ -0,0 +1,191 @@
+    >>> urng = np.random.default_rng()
+    >>> rng = NumericalInverseHermite(dist, order=5, random_state=urng)
+
+Higher orders result is fewer number of intervals:
```
Done.

tirthasheshpatel commented on doc/source/tutorial/stats/sampling_hinv.rst:
```diff
@@ -0,0 +1,191 @@
+Monte-Carlo simulation to estimate the u-error. By default, 100000 samples are
+used. To change this, you can pass the number of samples as an argument:
+
+    >>> rng1 = NumericalInverseHermite(dist, u_resolution=1.e-10)
```
Done.

chrisb83 commented on scipy/stats/_rvs_sampling.py:
```diff
@@ -171,451 +169,3 @@ def rvs_ratio_uniforms(pdf, umax, vmin, vmax, size=1, c=0, random_state=None):
-    ----------
-    intervals : int
-        The number of intervals of the interpolant.
-    midpoint_error : float
```
how do we proceed with the deprecation of `midpoint_error`?

tirthasheshpatel replied:
I will set it to `nan` with a deprecation warning on access as you suggested in https://github.com/scipy/scipy/pull/14703#issuecomment-921990877.

tirthasheshpatel replied:
Done and a test also added.

swallan replied:
Have you examined [the NumPy deprecation policy?](https://numpy.org/neps/nep-0023-backwards-compatibility.html) 

> removal of deprecated functionality ... shall be done after at least 2 releases assuming the current 6-monthly release cycle; if that changes, there shall be at least 1 year between deprecation and removal.

Similarly, the [SciPy deprecation policy](https://docs.scipy.org/doc/scipy/reference/dev/core-dev/index.html#deprecations) says that for deprivations there is a two release cycle before removal after proposing the change on the mailing list.

chrisb83 replied:
Thanks, that works. Let's remove the reference to 1.9.0 then. I suggest to mention the new method `u_error` in the warning message.

tirthasheshpatel replied:
You are right, @swallan. All the deprecations need 1 year of deprecation cycle. I will change all the warning messages accordingly.

tirthasheshpatel replied:
@chrisb83 I have added a reference to `u_error` in the deprecation warning of `midpoint_error`.

chrisb83 commented:
I had another look at the changes. The main open points seem to be:

- Matt to confirm if performance looks good from his point of view (from my point of view, a slightly slower setup is not a big problem as long as sampling is fast. the method is designed to generate large samples)
- deprecation of `tol` and `midpoint_error`

I think the feedback from @swallen has been addressed.

Apart from that, I did not note anything that raises concerns. `qrvs` has been added and tested. I think this addition is relevant for all inversion methods in UNU.RAN.

chrisb83 commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -1370,6 +1406,544 @@ cdef class NumericalInversePolynomial(Method):
+                  domain=None,
+                  order=3,
+                  u_resolution=1e-12,
+                  tol=None,
```
Btw: What is the default u_res in PINV? Is it aligned? Probably 1e-10 would be sufficient.

tirthasheshpatel replied:
UNU.RAN has default `u_resolution` set to `1e-10`. I had to change this because previously `NumericalInverseHermite` had 1e-12 default `u_resolution`. But I agree that `1e-12` is too small and `1e-10` would be sufficient. I think we can change this without having major backward compatibility issues. If the setup worked with `1e-12` then it should also work with `1e-10` (and would be faster).

tirthasheshpatel replied:
Changed the default to `1e-10`

swallan commented:
Something that I have been thinking about it the cost of the increased setup time vs the time gained from sample generation. 
![image](https://user-images.githubusercontent.com/44255917/135736882-17769a14-088c-41bd-a396-da539a909e85.png)

For stats distributions the total run time after including setup is only advantageous for more than `10e7` variates. I included this graph as an example but a lot of the distributions exhibit the same behavior. Do you think it would make sense to branch whether a user is providing a stats distribution since we know of the existing overhead that way we can still utilize the fast vectored code for them?

tirthasheshpatel commented:
> Do you think it would make sense to branch whether a user is providing a stats distribution since we know of the existing overhead that way we can still utilize the fast vectored code for them?

We could branch if we gave up the new parameters (e.g. `order`, `guide_factor`). Although this would be faster, we are also sacrificing consistency with other generators under the hood. Generally, we expect that the user would use these methods for non-standard distributions not present in stats. If we make it clear in the documentation that these generators don't integrate very well with stats distributions and should be used for non-standard custom distribution, we can probably get away with the slowdown. In a long term though, it would be nice to rewrite some of these UNU.RAN algorithms in Cython/C to make use of vectorization. But, for now, I think what we have is good enough.

chrisb83 commented:
@swallan Thanks for the analysis. It clearly shows that if an `rv_constinuous` object is used, the setup step is really slow. I hope we are clear in the documentation about this, I remember that we mentioned it in the tutorial. Because if one sets up a custom class for the distribution, the setup is pretty fast. I copied the results of the CI benchmark below (exception: if order=5 is used for sin2, it is quite slow, too). 

time_hinv_rvs:

|                      dist |           order            | time |
| --- | --- | --- |
|                  norm(0, 1)     |     3  |   20.3±0ms |
|                    norm(0, 1)   |     5  |   8.73±0ms |
|                       sin2          |   3    | 8.17±0ms |
|                       sin2          |   5    | 19.8±0ms |
|                      sin10         |    3   |  8.51±0ms |
|                      sin10         |    5   |  13.0±0ms |

hinv_setup:

|                      dist |           order            | time |
| --- | --- | --- |
|                    norm(0, 1)   |       3   |  14.6±0ms |
|                   norm(0, 1)    |      5  |   4.99±0ms |
|                       sin2          |   3   |  15.2±0ms |
|                       sin2          |   5   |  755±0ms  |
|                      sin10         |    3  |   45.8±0ms |
|                      sin10         |    5  |   18.2±0ms |

So from my point of view, if the method is used in the right way, it should produce a fast generator without an excessive setup.

Potentially we can also add support for `LowLevelCallable` at some point (as an alternative to rewriting the UNU.RAN code). We mentioned this already in Tirth's GSoC project summary

chrisb83 reviewed (approved):
All my comments were addressed. 

@swallan Could you also mark the comments you had as resolved if you are fine with the changes. Thanks!

swallan commented:
@chrisb83 => I think my comments are outdated so there isn't a resolve button. They are fixed, though. Thanks!

I am unsure if this can be merged in immediately because of the API changes. I alluded to it before, but maybe I wasn't clear. If the API is changing you should send an email to the mailing list proposing the change, and then once approved, first merge in code indicating the future deprecation. This could potentially be included as a "shadow" method, but the original features need to be present for a while (6 or more months) with deprecation warnings before the change can occur. I don't personally know the exact process but the folks on the mailing list can probably provide more definite insight. 

I think this PR looks good, but something else I am looking into is whether a custom `CubicHermiteSpline` would method (or an extension of the current one) could provide both fast setup time and fast sampling time.

tirthasheshpatel commented:
Thanks for the reviews, @swallan!

> I think my comments are outdated so there isn't a resolve button.

I have tried to resolve them. If you do want to discuss any of your comments further, feel free to comment there and I will unresolve.

> If the API is changing you should send an email to the mailing list proposing the change, and then once approved, first merge in code indicating the future deprecation. This could potentially be included as a "shadow" method, but the original features need to be present for a while (6 or more months) with deprecation warnings before the change can occur. I don't personally know the exact process but the folks on the mailing list can probably provide more definite insight.

I will send an email with a summary of changes and deprecations.

> but the original features need to be present for a while (6 or more months) with deprecation warnings before the change can occur.

My opinion on the backward-incompatible deprecation of `midpoint_error` in this PR: You are right that original features should be present for a while before changing or removing them. Theoretically, the `midpoint_error` deprecation and change is the only one that's breaking backward compatibility and holding us back from merging these changes. But practically, the removal of `midpoint_error` in a backward-incompatible way shouldn't make much difference. The main purpose of the method is to provide a fast method to generate random variates. I believe most use-cases would just initialize the class and call the `rvs` or `ppf` methods and rarely make use of attributes like `midpoint_error`. As this is new, I guess a few people might be using `NumericalInverseHermite` already, and even fewer accessing attributes like `midpoint_error`. Moreover, although `midpoint_error` is a good approximation of the maximum error, it doesn't guarantee that the highest error would be present at the midpoint of intervals. In this PR, we are providing a better method to approximate the maximum error which might be practically more useful. Overall, it shouldn't make much difference by changing and deprecating `midpoint_error`. I will run this through the mailing list to get others' opinions.

swallan commented:
@tirthasheshpatel @chrisb83 Thanks for sending the email -- I apologize that I haven't been more attentive. I investigated a few things on this PR like trying to implement the hermite splines as an alternative to make the setup faster -- though I didn't get too far. 

WRT a) I have never had to deprecate anything, so I am not an expert in this domain.  In the past it looks like the full functionality is maintained for a bit, gh-13774 from [1.6 release](https://docs.scipy.org/doc/scipy/reference/release.1.6.0.html#scipy-spatial-changes) still permitted input but began warning st users know it may not be permitted in the future. I'm not really sure what to do, hopefully someone with more expertise can weigh in, or there are more clear examples of this kind of API change in the deprecation sections of previous released.

b) I think that going from 12-10 more or less fine, I agree since you can't guarantee the values will be the same from python -> c anyways. Maintaining backwards compatibility works well too.

swallan commented:
> Also, there could be many more users by then

It is certainly something to consider that as this is a relatively fresh method, there may not be a ton of existing users and as time goes on there can only be more and more. It might not be _the same_ to report the new u error when midpoint error is accessed, but it could still report the tuple of the min and max or possibly a psuedo-replicate it with linearly spaced set of values between the points so it behaves similarly? I think that would be unconventional, but like you said, there aren't any especially good options.

chrisb83 commented:
> midpoint_error used to report an array of numerical errors at the midpoint of the intervals of the interpolating Hermite spline. While u_error returns the maximum error and mean absolute error at equidistant points on the spline. So, although u_error is very similar to midpoint_error, it doesn't give the same result (i.e. midpoint_error is an array of floats while u_error is a tuple of floats). 

That is correct: we will not get the same value but if we return `self.u-error()[0]`, this is conceptually the same error measure, just that the old version only looked at the midpoints to estimate the error as an approximation. Here is the description in the paper:

![image](https://user-images.githubusercontent.com/33071866/137442264-65484e0a-24f6-4d39-badf-856596fac11e.png)

That is why I think replacing the value can be seen as an improvement.

tupui commented:
> > and `tol` is missing in the docs
> 
> `tol` is in the [current docs](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.NumericalInverseHermite.html), so yes, the default accuracy is changing here, too. (The name is also changing.) But not a problem if the existing function is deprecated entirely.

He meant in this PR I suppose. Not the current state on master.

chrisb83 commented:
Hi all,

I just want to be sure that I am clear about next steps. Let me also try to address some of the concerns. The main ones seem to be

- deprecation of `tol` including change of the default
- deprecation of `midpoint_error`
- speed of the new method with scipy distributions
- number of parameters


**deprecation of `tol`:**
> I also noticed tol's default changed from 1e-12 to None, and tol is missing in the docs so not sure that changed behavior or not.

Correct, the impact on the behavior of the function should be minimal, see https://github.com/scipy/scipy/pull/14703#issuecomment-942981600

Samuel also seemed to be fine with the change of `tol` (https://github.com/scipy/scipy/pull/14703#issuecomment-942990309). If not, we can easily keep the current value 1e-12, this would just be a minor inconsistency compared to `NumericalInversePolynomial` but nothing to worry about. So I am sure we can find a solution for `tol`.

**deprecation of `midpoint_error`:**

I tried to explain my view in https://github.com/scipy/scipy/pull/14703#issuecomment-944037097 Just to reiterate: `midpoint_error` is just not the right measure of the u_error: by design, the midpoint error will be less than tol (that is the criterion of the algorithm) but the important question is whether the accuracy stays below `tol` at other points, too => that is what u_error will check. If that is not acceptable, I am fine.

**speed of the new method with scipy distributions**
Refering to Ralf's comment above, I think an option would be indeed to use `_pdf` and `_cdf` if `dist` is from the class `rv_continuous`. At the moment, we always access `dist.pdf`, but I guess this could be changed (either by creating a new dist object in NumericalInverseHermite with dist_new.pdf = dist._pdf etc or by adding an if ... else ...). @tirthasheshpatel what do you think about this?

one would need to check how this improves the speed (https://github.com/scipy/scipy/pull/14703#issuecomment-932849702)

**number of parameters**:
that is anyways under discussion in gh-14830 and I thought this is not a blocker

My main question is: assuming we get good results if we somehow rely on the private distribution methods, would the `midpoint_error` topic still be a blocker?

rgommers commented:
> My main question is: assuming we get good results if we somehow rely on the private distribution methods, would the `midpoint_error` topic still be a blocker?

It seems there's no argument about your statement that what it currently does isn't great. So it shouldn't be hard to resolve this one way or the other.

@swallan said: 

> In principle, we could also do that: we set `midpoint_error = self.u_error()` together with the warning message: the old `midpoint_error` is an approximation of the new one, so we would indeed return something better. Maybe that is the best option?

That seems reasonable to me. A comment on the warning: code should not give warnings for valid/correct usage, so maybe just note the behavior change in the release notes instead?

swallan commented:
@rgommers I think chris said that, but it's not ideal because
- midpoint error is a (large) numpy array of errors
- u-error is the minimum and maximum absolute errors 

So I don't think it's realistic to swap like that. 


> Didn't follow the "meaning" bit here

I was intending to reference that returning `NaN` from `midpoint_error` rather than the computed values just takes away any meaning from the attribute without deleting it.

chrisb83 commented:
> midpoint error is a (large) numpy array of errors

ok, this was pointed out before, I thought it is just the max. The documentation is wrong in that case: `midpoint_error, float The maximum u-error at an interpolant interval midpoint.` @mdhaber was it intended to return the array or should max be applied? It looks like this is a bug...

mdhaber commented:
Oops. max should be applied.

tirthasheshpatel commented:
> Not sure if this was already discussed somewhere, but is it not possible to call the private `_pdf`, `_cdf` directly? I'd expect it to be possible to bypass the expensive checks (at least for the default domain/loc/scale), or accumulate many scalar calls into a vectorized call.

Yes, we can wrap the distribution to use the private methods which are much faster. I tried it locally and it worked. All the tests with SciPy distributions passed (with an exception of two distributions). I have pushed this now. Let me know if this works for you, @swallan @rgommers @chrisb83. 

UNU.RAN failed with an error `CDF not increasing` for `invgauss` and `wald`. This happens because of inaccuracies at the tail. It did not happen for `NumericalInverseHermite` before because python implementation cut the boundaries off. So, this should be considered a bug in those distributions and not a bug in UNU.RAN. I have added these distributions in `fail_dist`.

<details>
<summary>Comparing benchmark on master and here</summary>

```
       before           after         ratio
     [27dd4b0b]       [1b474d74]
     <master>         <hinv-unuran>
!      51.7±0.4ms           failed      n/a  stats.NumericalInverseHermite.time_fni(['invgauss', (0.14546264555347513,)])
!        49.6±2ms           failed      n/a  stats.NumericalInverseHermite.time_fni(['wald', ()])
+      74.5±0.5ms       1.57±0.04s    21.00  stats.NumericalInverseHermite.time_fni(['pearson3', (0.1,)])
+      81.1±0.3ms       1.32±0.03s    16.33  stats.NumericalInverseHermite.time_fni(['ncx2', (21, 1.0560465975116415)])
+        64.2±6ms         984±20ms    15.32  stats.NumericalInverseHermite.time_fni(['crystalball', (2.0, 3.0)])
+         152±8ms       1.75±0.01s    11.49  stats.NumericalInverseHermite.time_fni(['kappa4', (0.1, 0.0)])
+      55.6±0.4ms         581±10ms    10.45  stats.NumericalInverseHermite.time_fni(['genextreme', (-0.1,)])
+        156±20ms       1.54±0.01s     9.84  stats.NumericalInverseHermite.time_fni(['kappa4', (-0.1, 0.1)])
+        170±10ms       1.59±0.05s     9.36  stats.NumericalInverseHermite.time_fni(['kappa4', (0.0, 0.0)])
+        163±10ms       1.49±0.01s     9.19  stats.NumericalInverseHermite.time_fni(['kappa4', (0.0, 0.1)])
+        73.1±5ms         667±10ms     9.12  stats.NumericalInverseHermite.time_fni(['vonmises_line', (3.99390425810714,)])
+        76.0±3ms         672±10ms     8.85  stats.NumericalInverseHermite.time_fni(['vonmises', (3.99390425810714,)])
+        40.0±3ms         340±10ms     8.51  stats.NumericalInverseHermite.time_fni(['burr', (10.5, 4.3)])
+      91.5±0.3ms          775±5ms     8.46  stats.NumericalInverseHermite.time_fni(['trapezoid', (0.2, 0.8)])
+      32.0±0.2ms          233±9ms     7.29  stats.NumericalInverseHermite.time_fni(['truncnorm', (-1.0978730080013919, 2.730675410903198)])
+        90.9±4ms         633±10ms     6.96  stats.NumericalInverseHermite.time_fni(['triang', (0.15785029824528218,)])
+        47.5±2ms         317±10ms     6.67  stats.NumericalInverseHermite.time_fni(['fisk', (3.085754862225318,)])
+        25.0±1ms          135±2ms     5.41  stats.NumericalInverseHermite.time_fni(['truncnorm', (0.1, 2.0)])
+      36.7±0.6ms          198±7ms     5.39  stats.NumericalInverseHermite.time_fni(['lognorm', (0.9536822696057533,)])
+        27.7±2ms          140±3ms     5.06  stats.NumericalInverseHermite.time_fni(['laplace', ()])
+      39.7±0.1ms        196±0.7ms     4.93  stats.NumericalInverseHermite.time_fni(['gilbrat', ()])
+     21.1±0.07ms         87.5±3ms     4.14  stats.NumericalInverseHermite.time_fni(['wrapcauchy', (0.031071279018614728,)])
+        38.5±3ms          121±6ms     3.15  stats.NumericalInverseHermite.time_fni(['laplace_asymmetric', (2,)])
+      31.0±0.9ms         94.6±3ms     3.05  stats.NumericalInverseHermite.time_fni(['johnsonsu', (2.554395574161155, 2.2482281679651965)])
+      42.8±0.1ms        122±0.9ms     2.86  stats.NumericalInverseHermite.time_fni(['genpareto', (0.1,)])
+        56.5±3ms          145±2ms     2.56  stats.NumericalInverseHermite.time_fni(['skewcauchy', (0.5,)])
+      46.2±0.1ms          114±7ms     2.46  stats.NumericalInverseHermite.time_fni(['loglaplace', (3.2505926592051435,)])
+      30.7±0.1ms         74.9±1ms     2.44  stats.NumericalInverseHermite.time_fni(['moyal', ()])
+        32.5±1ms         71.9±5ms     2.21  stats.NumericalInverseHermite.time_fni(['johnsonsb', (4.317267509914106, 3.1837781130785063)])
+     41.4±0.04ms         88.8±4ms     2.14  stats.NumericalInverseHermite.time_fni(['gengamma', (4.4162385429431925, 3.119309167924276)])
+      41.7±0.2ms         85.8±2ms     2.06  stats.NumericalInverseHermite.time_fni(['fatiguelife', (29,)])
+        50.4±2ms          104±2ms     2.06  stats.NumericalInverseHermite.time_fni(['dweibull', (2.0685080649914673,)])
+      47.4±0.1ms         87.0±3ms     1.84  stats.NumericalInverseHermite.time_fni(['gengamma', (4.4162385429431925, -3.119309167924276)])
+        31.4±2ms         56.5±3ms     1.80  stats.NumericalInverseHermite.time_fni(['exponweib', (2.8923945291034436, 1.9505288745913174)])
+        38.5±1ms         66.7±2ms     1.73  stats.NumericalInverseHermite.time_fni(['burr12', (10, 4)])
+      36.8±0.2ms         61.7±1ms     1.68  stats.NumericalInverseHermite.time_fni(['genlogistic', (0.41192440799679475,)])
+      44.1±0.8ms         70.5±2ms     1.60  stats.NumericalInverseHermite.time_fni(['t', (2.7433514990818093,)])
+      28.4±0.2ms       45.2±0.6ms     1.59  stats.NumericalInverseHermite.time_fni(['gumbel_r', ()])
+     32.2±0.09ms         50.3±1ms     1.56  stats.NumericalInverseHermite.time_fni(['invweibull', (10.58,)])
+      37.5±0.2ms       56.9±0.6ms     1.52  stats.NumericalInverseHermite.time_fni(['nakagami', (4.967379486666624,)])
+        52.2±4ms         75.7±8ms     1.45  stats.NumericalInverseHermite.time_fni(['recipinvgauss', (0.6300426780936912,)])
+      28.3±0.1ms       39.2±0.4ms     1.38  stats.NumericalInverseHermite.time_fni(['gumbel_l', ()])
+        43.9±4ms         60.8±1ms     1.38  stats.NumericalInverseHermite.time_fni(['exponnorm', (1.5,)])
+        35.5±3ms         48.9±1ms     1.38  stats.NumericalInverseHermite.time_fni(['exponpow', (2.697119160358469,)])
+      26.9±0.1ms         36.8±3ms     1.37  stats.NumericalInverseHermite.time_fni(['logistic', ()])
+        51.4±3ms       68.9±0.7ms     1.34  stats.NumericalInverseHermite.time_fni(['betaprime', (5, 6)])
+      37.5±0.2ms       49.8±0.6ms     1.33  stats.NumericalInverseHermite.time_fni(['mielke', (10.4, 4.6)])
+        41.4±2ms         53.9±5ms     1.30  stats.NumericalInverseHermite.time_fni(['chi', (78,)])
+        33.2±2ms       41.6±0.7ms     1.25  stats.NumericalInverseHermite.time_fni(['powerlognorm', (2.1413923530064087, 0.44639540782048337)])
+      28.2±0.1ms       35.3±0.4ms     1.25  stats.NumericalInverseHermite.time_fni(['hypsecant', ()])
+        40.8±4ms         49.7±2ms     1.22  stats.NumericalInverseHermite.time_fni(['alpha', (3.570477051665046,)])
+      34.2±0.4ms         40.9±3ms     1.19  stats.NumericalInverseHermite.time_fni(['powernorm', (4.445365225459078,)])
+     38.4±0.08ms       45.5±0.8ms     1.19  stats.NumericalInverseHermite.time_fni(['invgamma', (4.066899613699307,)])
+        58.1±4ms       67.1±0.9ms     1.15  stats.NumericalInverseHermite.time_fni(['levy', ()])
+     27.7±0.05ms       31.8±0.5ms     1.15  stats.NumericalInverseHermite.time_fni(['halflogistic', ()])
+      45.2±0.3ms       50.0±0.3ms     1.11  stats.NumericalInverseHermite.time_fni(['f', (29, 18)])
-        51.2±2ms       44.8±0.4ms     0.88  stats.NumericalInverseHermite.time_fni(['foldcauchy', (4.716467345583189,)])
-      27.0±0.1ms       21.1±0.3ms     0.78  stats.NumericalInverseHermite.time_fni(['halfnorm', ()])
-      47.2±0.5ms       36.8±0.4ms     0.78  stats.NumericalInverseHermite.time_fni(['foldnorm', (1.952125337355587,)])
-        45.9±2ms         35.3±3ms     0.77  stats.NumericalInverseHermite.time_fni(['rayleigh', ()])
-      70.6±0.1ms         53.6±4ms     0.76  stats.NumericalInverseHermite.time_fni(['gennorm', (1.2988442399460265,)])
-      26.5±0.1ms       19.5±0.5ms     0.74  stats.NumericalInverseHermite.time_fni(['gompertz', (0.9474371307510525,)])
-        43.2±2ms         31.7±2ms     0.73  stats.NumericalInverseHermite.time_fni(['anglit', ()])
-        40.8±2ms       29.9±0.8ms     0.73  stats.NumericalInverseHermite.time_fni(['cauchy', ()])
-     34.6±0.08ms       25.2±0.8ms     0.73  stats.NumericalInverseHermite.time_fni(['lomax', (1.8771398388773268,)])
-        19.9±1ms         14.2±1ms     0.71  stats.NumericalInverseHermite.time_fni(['truncexpon', (4.690772545681048,)])
-      53.7±0.9ms         37.3±2ms     0.69  stats.NumericalInverseHermite.time_fni(['weibull_min', (1.7866166930421596,)])
-        26.4±3ms       17.9±0.5ms     0.68  stats.NumericalInverseHermite.time_fni(['kstwobign', ()])
-      18.1±0.9ms       12.0±0.6ms     0.67  stats.NumericalInverseHermite.time_fni(['reciprocal', (0.01, 1.25)])
-      18.6±0.1ms      11.8±0.08ms     0.63  stats.NumericalInverseHermite.time_fni(['loguniform', (0.01, 1.25)])
-        36.1±4ms         22.5±1ms     0.62  stats.NumericalInverseHermite.time_fni(['kappa3', (1.0,)])
-      59.8±0.4ms         37.2±2ms     0.62  stats.NumericalInverseHermite.time_fni(['halfgennorm', (0.6748054997000371,)])
-     34.1±0.04ms       20.6±0.5ms     0.60  stats.NumericalInverseHermite.time_fni(['tukeylambda', (3.1321477856738267,)])
-      52.6±0.8ms       30.8±0.4ms     0.59  stats.NumericalInverseHermite.time_fni(['semicircular', ()])
-      32.4±0.6ms       18.7±0.6ms     0.58  stats.NumericalInverseHermite.time_fni(['arcsine', ()])
-     33.1±0.09ms         19.0±1ms     0.57  stats.NumericalInverseHermite.time_fni(['cosine', ()])
-      38.6±0.3ms       18.2±0.1ms     0.47  stats.NumericalInverseHermite.time_fni(['halfcauchy', ()])
-     6.59±0.02ms       2.88±0.1ms     0.44  stats.NumericalInverseHermite.time_fni(['uniform', ()])
-      8.44±0.6ms      3.55±0.01ms     0.42  stats.NumericalInverseHermite.time_fni(['bradford', (0.29891359763170633,)])
-      59.3±0.4ms       24.9±0.1ms     0.42  stats.NumericalInverseHermite.time_fni(['gamma', (1.9932305483800778,)])
-         104±9ms         35.8±2ms     0.35  stats.NumericalInverseHermite.time_fni(['rdist', (1.6,)])
-      56.2±0.5ms       14.6±0.6ms     0.26  stats.NumericalInverseHermite.time_fni(['genhalflogistic', (0.7727472780992932,)])
-      45.6±0.1ms       11.0±0.3ms     0.24  stats.NumericalInverseHermite.time_fni(['powerlaw', (1.659113328990585,)])
```

</details>

> I also noticed `tol`'s default changed from 1e-12 to None, and `tol` is missing in the docs so not sure that changed behavior or not.

`tol` has been deprecated and removed from documentation.

> Oops. max should be applied.

I think then changing `midpoint_error` to `u_error()[0]` would be considered a bug-fix and not a backward-incompatible change.

Do you agree we can proceed with this PR if we go with the above two changes, @swallan?

tirthasheshpatel commented:
I also added support for `loc` and `scale` passed with SciPy distribution. I will add tests for it later.

chrisb83 commented:
FYI: I opened an issue (gh-14875) for the midpoint_error topic.

@tirthasheshpatel thanks for adding better support for scipy distributions. I did not review the changes yet, just one question: how did you compute the timing results above? Is it the setup time?

chrisb83 commented:
I plan to merge this PR tomorrow. Please let me know if there are any concerns.

tirthasheshpatel commented on scipy/stats/_unuran/unuran_wrapper.pyx.templ:
```diff
@@ -257,6 +257,7 @@ cdef dict _unpack_dist(object dist, str dist_type, list meths,
                     (self.args, self.loc,
                      self.scale) = dist.dist._parse_args(*dist.args,
                                                          **dist.kwds)
+                    self.support = dist.support
```
One last change: Added the `support` method in the `wrap_dist` class so that methods like `DiscreteAliasUrn` can work without explicitly passing a domain parameter. Also added a test for it.

tirthasheshpatel commented:
@swallan Do the changes look good to you now? I and @chrisb83 hope to get this in before 1.8 branches but would appreciate your review/approval before we merge. Let us know if you have time to check out the changes!

chrisb83 commented:
No further feedback so far. I think we have addressed all points apart from the complex interface which we take up separately (https://github.com/scipy/scipy/pull/14703#issuecomment-954001442). I would finally merge this PR tomorrow morning (European time), so last chance to raise your concerns. Thanks!

chrisb83 commented:
Merged! Thanks a lot Tirth for your work on this. Thanks to all for the feedback!
